### PR TITLE
New pooled memory types

### PR DESCRIPTION
**Pull request by jhorv**
This PR adds 2 new types for using pooled memory:

- `MemoryOwner<T>` is a generic class for easy access to pooled memory.
- `SpanOwner<T>` is generic value type (`readonly ref struct`) that serves the same purpose in more constrained scopes, without requiring an additional allocation to manage it.

Both are intended for future optimization and to eventually replace the existing `ByteMemoryPool` and `ByteMemoryPool.ByteMemoryPoolBuffer` types in `Ryujinx.Common.Memory`, which have the following limitations:

- They only support `byte` buffers
- The `IMemoryOwner<byte>` implementation (the nested `ByteMemoryPoolBuffer` class) is private and returned as the interface. Usage as the concrete type would be faster.
- They don't have a value-type version, meaning their use requires allocating a `ByteMemoryPoolBuffer` instance.

This PR also uses the `SpanOwner` type to target a memory profiler hotspot, replacing a new array allocation with pooled memory in `Ryujinx.Graphics.Vulkan.MultiFenceHolder`. The table below shows the measured # of new `FenceHolder[]` allocations per second of gameplay (actual gameplay, past the title/intro/loading etc.).

### Before this PR (a.k.a., # of Allocations Eliminated by this PR)
| Game | Test Duration | Allocations | Allocations / second |
| -- | -- | -- | -- |
| Metroid Prime Remastered | 1 min 2.725 sec | 395,500 | **6,305** |
| TotK | 1 min 1.19 sec | 769,100 | **12,569** |

### CODE REVIEW

**Comment by gdkchan** [comment]
There are other custom pool implementations that we currently have, like https://github.com/Ryujinx/Ryujinx/blob/master/src/Ryujinx.Graphics.Vic/Image/BufferPool.cs
Do you think this would be eventually able to replace those?

File: src/Ryujinx.Common/Memory/MemoryOwner.cs [hunk 1]
```diff
@@ -0,0 +1,140 @@
...
+
+            return result;
+        }
+
+        /// <summary>
+        /// Gets the number of items in the current instance.
+        /// </summary>
+        public int Length
+        {
+            [MethodImpl(MethodImplOptions.AggressiveInlining)]
```
Comment (gdkchan) [inline src/Ryujinx.Common/Memory/MemoryOwner.cs:75]: I'm wondering why we need the aggressive inlining flag here, this is simple enough that I would expect the JIT to naturally inline it without the hint. Is it "just to be safe", or was it not inlining in some cases?

File: src/Ryujinx.Common/Memory/MemoryOwner.cs [hunk 2]
```diff
@@ -0,0 +1,140 @@
...
+            get
+            {
+                T[]? array = _array;
+
+                if (array is null)
+                {
+                    ThrowObjectDisposedException();
+                }
+
+                ref T firstElementRef = ref MemoryMarshal.GetArrayDataReference(array);
```
Comment (gdkchan) [inline src/Ryujinx.Common/Memory/MemoryOwner.cs:114]: I'm a little confused. Why is this necessary, and it can't just do `return array;` directly? It is possible to implicit cast an array to span.

**Comment by jhorv** [comment]
> There are other custom pool implementations that we currently have, like https://github.com/Ryujinx/Ryujinx/blob/master/src/Ryujinx.Graphics.Vic/Image/BufferPool.cs Do you think this would be eventually able to replace those?

They could be replaced, but I think it'd be wise to study them for possible behavior/perf differences. If any of them wrap ArrayPool, those could be drop-in replaced by these new ones without much concern. But any that have a custom algorithm (like `BufferPool` you mention) may have significantly different perf characteristics that we should probably measure before replacing.

File: src/Ryujinx.Common/Memory/MemoryOwner.cs (hunk 1 above)
Comment (jhorv) [reply src/Ryujinx.Common/Memory/MemoryOwner.cs:75]: It is "just to be safe".

File: src/Ryujinx.Common/Memory/MemoryOwner.cs (hunk 2 above)
Comment (jhorv) [reply src/Ryujinx.Common/Memory/MemoryOwner.cs:114]: If you [look at the code](https://github.com/dotnet/runtime/blob/84885d7a878503c350e136f4297bfa234cf66f73/src/libraries/System.Private.CoreLib/src/System/Span.cs#L42) for the relevant `public Span(T[]? array)` constructor, which is called by the [implicit conversion operator](https://github.com/dotnet/runtime/blob/84885d7a878503c350e136f4297bfa234cf66f73/src/libraries/System.Private.CoreLib/src/System/Span.cs#L211), you'll see it does a bunch of checks: for nullness, for type covariance, etc.

This is a "high-performance" approach that avoids all that.

**Comment by gdkchan** [comment]
> > There are other custom pool implementations that we currently have, like https://github.com/Ryujinx/Ryujinx/blob/master/src/Ryujinx.Graphics.Vic/Image/BufferPool.cs Do you think this would be eventually able to replace those?
> 
> They could be replaced, but I think it'd be wise to study them for possible behavior/perf differences. If any of them wrap ArrayPool, those could be drop-in replaced by these new ones without much concern. But any that have a custom algorithm (like `BufferPool` you mention) may have significantly different perf characteristics that we should probably measure before replacing.

I think `ArrayPool` originally had a pretty low size cap before, low enough that if I tried to rent those buffers using this approach it would not be actually pooled and would just allocate a new array every time. I think around .NET 7 they increased it, so probably should not be a problem anymore. But yea, if its replaced in the future, it will need to be tested carefully. That specific use case is somewhat special since VIC usually allocates a few buffer for the video frames when the decoding starts, and keeps using them until decoding ends. So the lifetime of those buffers is pretty well defined in those cases.

**Review by gdkchan — APPROVED** [review]
lgtm, thanks.

**Review by riperiperi — APPROVED** [review]
Looks good, sorry for the late review!
